### ISSUE TITLE
Exception Handler called twice if controller doesn't exist

### ISSUE THREAD

**Issue by rafee12** [laravel/framework#17007]
- Laravel Version: 5.3.28
- PHP Version: 5.6.23

### Description:
exception handler called twice if the controller doesn't exist so it will render the error page or custom error page twice. It looks like the problem with \Illuminate\Routing\Route::gatherMiddleware(), in 5.2 the method \Illuminate\Routing\Router::gatherRouteMiddlewares() only gather \Illuminate\Routing\Route::middleware() but in 5.3 they gather \Illuminate\Routing\Route::middleware() and \Illuminate\Routing\Route::controllerMiddleware() using \Illuminate\Routing\Route::gatherMiddleware() method

\Illuminate\Routing\Route::controllerMiddleware() will check the controller existence again with ReflectionClass, and because the controller doesn't exist, it will throw ReflectionException again.

### Steps To Reproduce:

1.  just create a route with a controller that doesn't exist and open it

**Comment by GrahamCampbell** [comment]
That is unfortunate. :/

**Comment by srmklive** [comment]
I can't replicate the issue on an empty resource controller as shown in the following screenshot:

![_-_2016-12-28_15 31 02](https://cloud.githubusercontent.com/assets/839335/21519973/bde2c374-cd12-11e6-9860-89b0e31f9f4e.png)

But this issue does occur if i create a resource controller and goto a route that doesn't exists within a controller.

**Comment by rafee12** [comment]
@srmklive are you sure you are using 5.3? it's look like 5.2, the problem only occur on 5.3.

**Comment by GrahamCampbell** [comment]
This is happening because in 5.3, the response from an exception goes back out the middleware, but in 5.2, it did not.

**Comment by srmklive** [comment]
@rafee12 Yes i am using 5.3.

**Comment by bancur** [comment]
I've made a pull request that fixes this issue:

https://github.com/laravel/framework/pull/17061

It fixes it on my setup - can you confirm that it does the same for you?

**Comment by GrahamCampbell** [comment]
Hopefully this is a better fix: https://github.com/laravel/framework/pull/17064. We can't accept #17061 because it just blindly silences bad controller middleware, when what we really want is to just not have the same error thrown more than once.

**Comment by MrMashy** [comment]
+1, Have we got a fix yet?

**Comment by GrahamCampbell** [comment]
Yes. This fixed in L5.4. A stable release is likely to be coming later this month. I don't have a release date to give out though, sorry.

**Comment by MrMashy** [comment]
Sweet, I'll be sure to update my application to that.

**Comment by MrMashy** [comment]
In the meantime, could you give me a code snippet to fix my current project?

**Comment by GrahamCampbell** [comment]
It requires making changes to the router.

**Comment by GrahamCampbell** [comment]
To patch 5.4, in `Route.php`, add:

```php
    /**
     * The computed gathered middleware.
     *
     * @var array|null
     */
    protected $computedMiddleware;
```

above 

```php
    /**
     * The compiled version of the route.
     *
     * @var \Symfony\Component\Routing\CompiledRoute
     */
    protected $compiled;
```

and then replace the gather middleware method with:

```php
    /**
     * Get all middleware, including the ones from the controller.
     *
     * @return array
     */
    public function gatherMiddleware()
    {
        if (! is_null($this->computedMiddleware)) {
            return $this->computedMiddleware;
        }

        $this->computedMiddleware = [];

        return $this->computedMiddleware = array_unique(array_merge(
            $this->middleware(), $this->controllerMiddleware()
        ), SORT_REGULAR);
    }
```

**Comment by MrMashy** [comment]
Thanks, I also solved the problem by removing `$kernel->terminate($request, $response);` from index.php

Obviously probably not a good idea however.

**Comment by GrahamCampbell** [comment]
> Obviously probably not a good idea however.

Yes, that won't solve the problem. That'll break all terminating middleware.

**Comment by saidbakr** [comment]
I have 5.4 based app and it occurs with redis.
